File: modules/session-audio-engine/ios/VoiceCapturePipeline.swift
import Foundation

final class VoiceCapturePipeline {
  var onCaptured: ((NativeCapturedSegment) -> Void)?
  var onFailure: ((Error) -> Void)?

  private let configuration: NativeSessionConfiguration
  private let store: PendingCaptureStore
  private var pendingSamples: [Int16] = []
  private var ringBuffer: [Int16] = []
  private var segmentSamples: [Int16] = []
  private var aboveMs: Int64 = 0
  private var belowMs: Int64 = 0
  private var voiceActive = false
  private var speechStartMs: Int64 = 0
  private var phase = "learning"
  private var cycle = 1

  init(configuration: NativeSessionConfiguration, store: PendingCaptureStore) {
    self.configuration = configuration
    self.store = store
  }

  func process(samples: [Int16], captureAllowed: Bool, phase: String, cycle: Int) {
    guard captureAllowed else {
      flush()
      ringBuffer.removeAll(keepingCapacity: true)
      aboveMs = 0
      return
    }
    self.phase = phase
    self.cycle = cycle
    pendingSamples.append(contentsOf: samples)
    while pendingSamples.count >= 1_600 {
      let chunk = Array(pendingSamples.prefix(1_600))
      pendingSamples.removeFirst(1_600)
      processChunk(chunk)
    }
  }

  func flush() {
    if voiceActive { finalizeSegment(releaseTailMs: belowMs) }
    resetDetection()
  }

  private func processChunk(_ chunk: [Int16]) {
    let normalized = normalizedLevel(chunk)
    let chunkMs: Int64 = 100
    if !voiceActive {
      ringBuffer.append(contentsOf: chunk)
      let preRollSamples = Int(configuration.vad.preRollMs) * 16
      if ringBuffer.count > preRollSamples {
        ringBuffer.removeFirst(ringBuffer.count - preRollSamples)
      }
      aboveMs = normalized > configuration.vad.threshold ? aboveMs + chunkMs : 0
      if aboveMs >= configuration.vad.sustainMs {
        voiceActive = true
        segmentSamples = ringBuffer
        speechStartMs = max(0, Int64(segmentSamples.count) / 16 - aboveMs)
        belowMs = 0
      }
      return
    }

    segmentSamples.append(contentsOf: chunk)
    belowMs = normalized <= configuration.vad.threshold ? belowMs + chunkMs : 0
    let durationMs = Int64(segmentSamples.count) / 16
    if belowMs >= configuration.vad.releaseMs || durationMs >= configuration.vad.maxSegmentMs {
      finalizeSegment(releaseTailMs: belowMs)
      resetDetection()
    }
  }

  private func finalizeSegment(releaseTailMs: Int64) {
    guard !segmentSamples.isEmpty else { return }
    let durationMs = Int64(segmentSamples.count) / 16
    let speechEndMs = max(speechStartMs, durationMs - releaseTailMs)
    do {
      let segment = try store.store(
        samples: segmentSamples,
        configuration: configuration,
        phase: phase,
        cycle: cycle,
        speechStartMs: speechStartMs,
        speechEndMs: speechEndMs
      )
      onCaptured?(segment)
    } catch {
      onFailure?(error)
    }
  }

  private func resetDetection() {
    voiceActive = false
    aboveMs = 0
    belowMs = 0
    speechStartMs = 0
    segmentSamples.removeAll(keepingCapacity: true)
    ringBuffer.removeAll(keepingCapacity: true)
  }

  private func normalizedLevel(_ samples: [Int16]) -> Double {
    let meanSquare = samples.reduce(0.0) { partial, sample in
      let normalized = Double(sample) / Double(Int16.max)
      return partial + normalized * normalized
    } / Double(max(1, samples.count))
    let db = meanSquare > 0 ? 20 * log10(sqrt(meanSquare)) : configuration.vad.dbFloor
    let range = configuration.vad.dbCeil - configuration.vad.dbFloor
    guard range > 0 else { return 0 }
    return max(0, min(1, (db - configuration.vad.dbFloor) / range))
  }
}

File: modules/session-audio-engine/android/src/main/java/com/joynnovate/buddybird/sessionaudio/VoiceCapturePipeline.kt
package com.joynnovate.buddybird.sessionaudio

import kotlin.math.log10
import kotlin.math.sqrt

class VoiceCapturePipeline(
  private val configuration: NativeSessionConfiguration,
  private val store: PendingCaptureStore,
) {
  var onCaptured: ((NativeCapturedSegment) -> Unit)? = null
  var onFailure: ((Throwable) -> Unit)? = null

  private val pendingSamples = ArrayList<Short>()
  private val ringBuffer = ArrayList<Short>()
  private val segmentSamples = ArrayList<Short>()
  private var aboveMs = 0L
  private var belowMs = 0L
  private var voiceActive = false
  private var speechStartMs = 0L
  private var phase = "learning"
  private var cycle = 1

  @Synchronized
  fun process(samples: ShortArray, count: Int, captureAllowed: Boolean, phase: String, cycle: Int) {
    if (!captureAllowed) {
      flush()
      ringBuffer.clear()
      aboveMs = 0
      return
    }
    this.phase = phase
    this.cycle = cycle
    for (index in 0 until count) pendingSamples += samples[index]
    while (pendingSamples.size >= 1_600) {
      val chunk = ShortArray(1_600) { pendingSamples[it] }
      pendingSamples.subList(0, 1_600).clear()
      processChunk(chunk)
    }
  }

  @Synchronized
  fun flush() {
    if (voiceActive) finalizeSegment(belowMs)
    resetDetection()
  }

  private fun processChunk(chunk: ShortArray) {
    val normalized = normalizedLevel(chunk)
    if (!voiceActive) {
      chunk.forEach { ringBuffer += it }
      val preRollSamples = configuration.vad.preRollMs.toInt() * 16
      if (ringBuffer.size > preRollSamples) ringBuffer.subList(0, ringBuffer.size - preRollSamples).clear()
      aboveMs = if (normalized > configuration.vad.threshold) aboveMs + 100 else 0
      if (aboveMs >= configuration.vad.sustainMs) {
        voiceActive = true
        segmentSamples.addAll(ringBuffer)
        speechStartMs = (segmentSamples.size.toLong() / 16 - aboveMs).coerceAtLeast(0)
        belowMs = 0
      }
      return
    }

    chunk.forEach { segmentSamples += it }
    belowMs = if (normalized <= configuration.vad.threshold) belowMs + 100 else 0
    val durationMs = segmentSamples.size.toLong() / 16
    if (belowMs >= configuration.vad.releaseMs || durationMs >= configuration.vad.maxSegmentMs) {
      finalizeSegment(belowMs)
      resetDetection()
    }
  }

  private fun finalizeSegment(releaseTailMs: Long) {
    if (segmentSamples.isEmpty()) return
    val samples = ShortArray(segmentSamples.size) { segmentSamples[it] }
    val durationMs = samples.size.toLong() / 16
    runCatching {
      store.store(
        samples = samples,
        configuration = configuration,
        phase = phase,
        cycle = cycle,
        speechStartMs = speechStartMs,
        speechEndMs = (durationMs - releaseTailMs).coerceAtLeast(speechStartMs),
      )
    }.onSuccess { onCaptured?.invoke(it) }.onFailure { onFailure?.invoke(it) }
  }

  private fun resetDetection() {
    voiceActive = false
    aboveMs = 0
    belowMs = 0
    speechStartMs = 0
    segmentSamples.clear()
    ringBuffer.clear()
    pendingSamples.clear()
  }

  private fun normalizedLevel(samples: ShortArray): Double {
    val meanSquare = samples.sumOf {
      val normalized = it.toDouble() / Short.MAX_VALUE
      normalized * normalized
    } / samples.size.coerceAtLeast(1)
    val db = if (meanSquare > 0) 20 * log10(sqrt(meanSquare)) else configuration.vad.dbFloor
    val range = configuration.vad.dbCeil - configuration.vad.dbFloor
    return if (range > 0) ((db - configuration.vad.dbFloor) / range).coerceIn(0.0, 1.0) else 0.0
  }
}
